The problem is in exception handler return expander (expand_eh_return in
except.c) that clears EH_RETURN_STACKADJ_RTX. Optimizer passes then push this
dangling insn around and sometimes it gets pushed into exit block, where it
confuses create_pre_exit (). Later, this instruction is cleared anyway, as it is
in fact not used anywhere. This looks like a generic bug to me.

Thanks in advance,
Uros.

--- gcc-4.1-20050818T1605UTC/gcc/except.c.orig	2005-08-18 16:00:40.000000000 +0000
+++ gcc-4.1-20050818T1605UTC/gcc/except.c	2005-08-18 16:28:54.000000000 +0000
@@ -2864,10 +2864,6 @@
 
   current_function_calls_eh_return = 1;
 
-#ifdef EH_RETURN_STACKADJ_RTX
-  emit_move_insn (EH_RETURN_STACKADJ_RTX, const0_rtx);
-#endif
-
   around_label = gen_label_rtx ();
   emit_jump (around_label);
 
